scipy/optimize/optimize.py

```diff
@@ -776,12 +777,13 @@ def _line_search_wolfe12(f, fprime, xk, pk, gfk, old_fval, old_old_fval,
 
     if ret[0] is None:
         # line search failed: try different one.
-        with warnings.catch_warnings():
-            warnings.simplefilter('ignore', LineSearchWarning)
-            kwargs2 = {}
-            for key in ('c1', 'c2', 'amax'):
-                if key in kwargs:
-                    kwargs2[key] = kwargs[key]
+        kwargs2 = {}
+        for key in ('c1', 'c2', 'amax'):
+            if key in kwargs:
+                kwargs2[key] = kwargs[key]
+        with suppress_warnings() as sup:
+            sup.filter(LineSearchWarning,
+                       "The line search algorithm did not converge")
```

Comment:
Yeah, even if it does not do the worst stuff here, I remembered the thread safety issue.... Now catch_warnings is not thread safe as well, but at least it can't blow up due to a bad internal state, suppress_warnings can very much do that.